File: Graphs/distance_k.py
from collections import deque
from typing import List

from BinaryTrees.tree_node import TreeNode


class DistanceK:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        # Find target and assign parent pointers
        queue = deque([root])
        while queue:
            current = queue.popleft()
            if current == target:
                break

            if current.left:
                current.left.parent = current
                queue.append(current.left)
            if current.right:
                current.right.parent = current
                queue.append(current.right)

        # Use BFS with target as origin
        queue = deque([target])
        seen = set([target])
        while queue and k > 0:
            level_size = len(queue)
            for _ in range(level_size):
                current = queue.popleft()
                if current.left and current.left not in seen:
                    seen.add(current.left)
                    queue.append(current.left)
                if current.right and current.right not in seen:
                    seen.add(current.right)
                    queue.append(current.right)
                if hasattr(current, "parent") and current.parent not in seen:
                    seen.add(current.parent)
                    queue.append(current.parent)
            k -= 1

        return [node.val for node in queue]

File: Graphs/DistanceK.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/*Given the root of a binary tree, the value of a target node target, and an integer k, 
return an array of the values of all nodes that have a distance k from the target node.

You can return the answer in any order. */

class DistanceK {
    Map<TreeNode, TreeNode> parents;
    Set<TreeNode> seen;
    int k;

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        parents = new HashMap<>();
        seen = new HashSet<>();
        this.k = k;

        setParents(root, null);
        return getNodesAtKDistance(target);
    }

    private void setParents(TreeNode node, TreeNode parent) {
        if (node == null) {
            return;
        }

        parents.put(node, parent);

        setParents(node.left, node);
        setParents(node.right, node);
    }

    private List<Integer> getNodesAtKDistance(TreeNode startingNode) {
        seen.add(startingNode);
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(startingNode);

        while (k > 0 && !queue.isEmpty()) {
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                TreeNode currentNode = queue.removeFirst();

                for (TreeNode neighbor : new TreeNode[] { currentNode.left, currentNode.right,
                        parents.get(currentNode) }) {
                    if (neighbor != null && !seen.contains(neighbor)) {
                        seen.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }

            k--;
        }

        List<Integer> kDistanceNodes = queue.stream().map(node -> node.val).toList();

        return kDistanceNodes;
    }
}

// Time Complexity: O(n) where n is the number of nodes in the tree. We visit
// each node at most twice (once to set parents and once to find nodes at
// distance k).
// Space Complexity: O(n) for the parents map, seen set, and queue in the worst
// case when the tree is skewed. In a balanced tree, the space complexity would
// be O(h) where h is the height of the tree.